### ISSUE TITLE
Sampling seems to be broken

### ISSUE THREAD

**Issue by damianoazzolini** [David-Tuc/pasocs#4]
Hi,
I'm trying to use sampling. However, I think pasocs does not perform sampling even if I use the required parameters.
For example (call this file `test.lp`):
```
0.5::a. 0.5::b. 0.5::c. 0.5::d.
0.5::e. 0.5::f. 0.5::g. 0.5::h.
0.5::i. 0.5::l. 0.5::m. 0.5::n.
0.5::o. 0.5::p.

q:-a. q:-b. q:-c.
q:-d. q:-e. q:-f.
q:-g. q:-h. q:-i.
q:-l. q:-m. q:-n.
q:-o. q:-p.
```
Using: `./pasocs test.lp -sm 1 -n_max 10 -n_min 1 -v 1 --query q` I get
```
Output:
[{q - Lower=0.999939+/-0.000000 - Upper=0.999939+/-0.000000 - N Samples=0}]
Total time: 7.94713s
Solving time:7.86204s
```
The probability is correct, but `N Samples` is 0 and the execution time is the same required to perform exact inference (`/pasocs test.lp --query q`). Moreover, the execution time does not change with a different number of samples.

**Comment by DvdTky** [comment]
Hi, 

Thank you for posting this issue. I have just fixed the bug in the code (the solving mode was not taken into account) and updated the binaries. 

Please let me know if the issue persists.

**Comment by damianoazzolini** [comment]
Thanks, the program now works fine.
Just another question: how can I set the number of samples and impose that the program actually perform that number of samples? For instance (consider this program saved into a file called `test.lp`),
```
0.5::a. 0.5::b. 0.5::c. 0.5::d.
q:-a. q:-b. q:-c. q:-d.
```
with
`./pasocs test.lp -sm 3 -n_max 10000 -n_min 10000 -v 1 --query "q|b" -ut -1 -p 300 -b 0`
i get
`[{q|b:true - Lower=1.000000+/-0.000000 - Upper=1.000000+/-0.000000 - N Samples=520},]`
so the number of samples is 520, and not 10000 as I specified.
I know that these are sufficient to compute the probability, but is there a way to impose the number of samples and ignoring all the stopping criteria?
Or Gibbs and Mh do not support sampling with evidence? Without evidence, this works fine.
Thanks

**Comment by DvdTky** [comment]
Yes sorry the criteria for sampling can be a bit confusing. The idea is that when doing sampling with evidence, some samples do not satisfy the evidence and are not taken into account for a query. 
What happen here, is that the system sampled 10000 worlds but only had 520 that satisfied the query's evidence.

This is because you set -n_max to 10000, setting the hard maximum amount of sample it is allowed to take.
Now if you want 10000 samples for your query, you should not set -n_max to -1. This will make that there is no hard limit on the number of samples, and it will stop once all your queries have reached the "soft" limit of having -n_min samples each.

So running:
./pasocs test.lp -sm 3 -n_max -1 -n_min 10000 -v 1 --query "q|b" -ut -1 -p 300 -b 0

should give you what you are looking for. Note that the actual number of samples for your query will probably be above 10000 as it computes samples in batches. To have exactly 10000, you should set -sb to 1, but this will make for a longer running time when using multiple CPUs.

With:
./pasocs test.lp -sm 3 -n_max -1 -n_min 10000 -v 1 --query "q|b" -ut -1 -p 300 -b 0 -sb 1

Output:
[{q|b:true - Lower=1.000000+/-0.000000 - Upper=1.000000+/-0.000000 - N Samples=10000},
]
Total time: 2.53443s
Solving time:2.52167s

**Comment by damianoazzolini** [comment]
Ok now it's clear! Thanks for the quick response.
